### PR TITLE
c7n.k8s - support custom resource definitions

### PR DESCRIPTION
**Pull request by thisisshi**
- adds support for custom resource definitions
- adds support for docstrings for resources (custom resources require some additional fields that are passed in via the `query` attribute in the policy)
- adds tests
- add generic patch action

```
$ custodian schema k8s.custom-namespaced-resource
Help
----

Query Custom Resource Definitions

Custom resources require query to be defined with the group,
version, and plural values from the resource definition

policies:
  - name: custom-resource
    resource: k8s.custom-namespaced-resource
    query:
      - group: stable.example.com
        version: v1
        plural: crontabs

k8s.custom-namespaced-resource:
  actions:
  - notify
  filters:
  - and
  - event
  - not
  - or
  - value
```

### CODE REVIEW

**Comment by thisisshi** [comment]
hrm.. not sure why the test is failing in 2.7, I'm asserting that it should raise the PolicyValidationError which it does

File: c7n/commands.py [hunk 1]
```diff
@@ -293,12 +293,14 @@ def logs(options, policies):
...
 
 
-def _schema_get_docstring(starting_class):
+def _schema_get_docstring(starting_class, base_only=False):
     """ Given a class, return its docstring.
 
     If no docstring is present for the class, search base classes in MRO for a
     docstring.
     """
+    if base_only:
```
Comment (kapilt) [inline c7n/commands.py]: why? just put a doc string in the class

File: c7n/policy.py [hunk 2]
```diff
@@ -826,6 +826,9 @@ def validate(self):
                 "Invalid Execution mode in policy %s" % (self.data,))
         m.validate()
         self.validate_policy_start_stop()
+        # support validation on k8s custom resource types
+        if hasattr(self.resource_manager, 'validate'):
```
Comment (kapilt) [inline c7n/policy.py]: no hasattr please, extend the interface of resource manager with validate, and a doc string to clarify its what its not doing (mode, filters, actions)

File: tools/c7n_kube/c7n_kube/resources/crd.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            version: v1
+            plural: crontabs
+    """
+    class resource_type(CustomTypeInfo):
+        delete = "delete_namespaced_custom_object"
+        patch = "patch_namespaced_custom_object"
+
+
+@resources.register('custom-cluster-resource')
+class CustomResourceDefinition(CustomResourceQueryManager):
```
Comment (kapilt) [inline tools/c7n_kube/c7n_kube/resources/crd.py]: are there custom resource that live outside of namespaces?

**Comment by kapilt** [comment]
could you clarify intent, this is to find custom resource definitions (aka classes) not instances of the custom resource? Worth noting in the doc string.

File: tools/c7n_kube/c7n_kube/resources/crd.py (hunk 3 above)
Comment (thisisshi) [reply tools/c7n_kube/c7n_kube/resources/crd.py]: yup, you can define custom resource types to be either namespaced to either the cluster or a namespace

**Comment by thisisshi** [comment]
@kapilt instance of a custom resource definition, not the definition itself

This will probably also come with a generic patch action to acton on those resources as well

File: c7n/policy.py (hunk 2 above)
Comment (thisisshi) [reply c7n/policy.py]: sounds good

File: tools/c7n_kube/tests/test_custom_resource.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                        'version': 'v1',
+                        'plural': 'crontabscluster'
+                    }
+                ]
+            },
+            session_factory=factory
+        )
+
+        resources = policy.run()
+        self.assertTrue(resources)
```
Comment (kapilt) [inline tools/c7n_kube/tests/test_custom_resource.py]: any attribute validation that can be added to these assertions?

**Comment by thisisshi** [comment]
i'll add one more test on the patch action too

**Comment by kapilt** [comment]
doh, sorry missed your last comment

**Pull request merged**